### PR TITLE
Add DaemonSet module

### PR DESCRIPTION
**Pull request by slawqo**

### CODE REVIEW

File: modules/common/daemonset/daemonset.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// GetDaemonSet - get the daemonset object.
+func (d *DaemonSet) GetDaemonSet() appsv1.DaemonSet {
+	return *d.daemonset
+}
+
+// SetTimeout defines the duration used for requeueing while waiting for the daemonset
+// to finish.
+func (d *DaemonSet) SetTimeout(timeout time.Duration) {
```
Comment (gibizer) [inline modules/common/daemonset/daemonset.go:115]: I would remove this and use the NewDaemonSet with a time.Duration parameter instead.

File: modules/common/daemonset/daemonset.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	d.timeout = timeout
+}
+
+// GetDaemonSetWithName func
+func GetDaemonSetWithName(
```
Comment (gibizer) [inline modules/common/daemonset/daemonset.go:120]: Does this need to be an exported call? If not, then please rename it to getDaemonSetWithName to make it private.

**Review by gibizer — CHANGES_REQUESTED** [review]

File: modules/common/daemonset/daemonset.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		err := controllerutil.SetControllerReference(h.GetBeforeObject(), daemonset, h.GetScheme())
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			h.GetLogger().Info(fmt.Sprintf("DaemonSet %s not found, reconcile in %s", daemonset.Name, d.timeout))
```
[the remark this answers is not in the document]
Comment (stuggi) [reply modules/common/daemonset/daemonset.go:76]: yes we should, so probably good to do it here right from the start. I have a task on my list to cleanup lib-common to do it where it is not used.

File: modules/common/daemonset/daemonset.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+			Name:      d.daemonset.Name,
+			Namespace: d.daemonset.Namespace,
+		},
+	}
+
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), daemonset, func() error {
+		// Daemonset selector is immutable so we set this value only if
+		// a new object is going to be created
+		if daemonset.ObjectMeta.CreationTimestamp.IsZero() {
+			daemonset.Spec.Selector = d.daemonset.Spec.Selector
```
Comment (gibizer) [inline modules/common/daemonset/daemonset.go:61]: Would it be simpler and still correct to fill daemonset.Spec.Selector at L54 above? I think (but not 100% sure) that would mean Selector is only set at Create but not during Patch so we could remove the IsZero conditional.

File: modules/common/daemonset/daemonset.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+			return ctrl.Result{RequeueAfter: d.timeout}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if op != controllerutil.OperationResultNone {
+		h.GetLogger().Info(fmt.Sprintf("DaemonSet %s - %s", daemonset.Name, op))
+	}
+
+	// update the daemonset object of the daemonset type
+	d.daemonset, err = GetDaemonSetWithName(ctx, h, daemonset.GetName(), daemonset.GetNamespace())
```
[the remark this answers is not in the document]
Comment (stuggi) [reply modules/common/daemonset/daemonset.go:86]: no it does not update the passed in object, we have to get the actual one from the env

[the remark this answers is not in the document]
Comment (gibizer) [reply modules/common/daemonset/daemonset.go:86]: Thanks for clarifying it. 
It is sad inconsistency that the Client().Patch() does update the passed in object but CreateOrPatch() doesn't.

File: modules/common/daemonset/daemonset.go (hunk 4 above)
Comment (stuggi) [reply modules/common/daemonset/daemonset.go:61]: not sure, we could test that, I just took this approach initially from the controllerutil example https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/example_test.go#L44

File: modules/common/daemonset/daemonset.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+// NewDaemonSet returns an initialized DaemonSet
+func NewDaemonSet(
+	daemonset *appsv1.DaemonSet,
+	timeout time.Duration,
+) *DaemonSet {
+	return &DaemonSet{
+		daemonset: daemonset,
+		timeout:   timeout * time.Second,
```
Comment (gibizer) [inline modules/common/daemonset/daemonset.go:41]: this should be
```
timeout: timeout
```
The time.Duration type allows any precision not just Seconds.

**Review by gibizer — CHANGES_REQUESTED** [review]
Thanks for the fixes. I have some followup inline.

File: modules/common/daemonset/daemonset.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		err = fmt.Errorf("error deleting daemonset %s: %v", d.daemonset.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetDaemonSet - get the daemonset object.
+func (d *DaemonSet) GetDaemonSet() appsv1.DaemonSet {
```
[the remark this answers is not in the document]
Comment (abays) [reply modules/common/daemonset/daemonset.go:112]: I think one possible use would be similar to what we do here in Cinder with `GetDeployment()` (which follows a similar pattern in terms of how `Deployment` is wrapped in `lib-common`): https://github.com/openstack-k8s-operators/cinder-operator/blob/60bdac1a3f2db7cd1ea1e8a43a3309bc3b4209e7/controllers/cinderapi_controller.go#L653

[the remark this answers is not in the document]
Comment (gibizer) [reply modules/common/daemonset/daemonset.go:112]: Ahh I see. So we need to access appsv1.DaemonSet from the caller as the caller needs to know when the DeamonSet is healthy / ready. OK. Then I think we can keep this as is.

**Review by gibizer — APPROVED** [review]
Looks good to me thanks.

**Review by stuggi — APPROVED** [review]
/lgtm

**Pull request merged**